<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · PanGraph.jl</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>PanGraph.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Introduction"><span>Introduction</span></a></li><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#User&#39;s-Guide"><span>User&#39;s Guide</span></a></li><li><a class="tocitem" href="#Citing-PanGraph"><span>Citing PanGraph</span></a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="lib/align/">Alignment</a></li><li><a class="tocitem" href="lib/block/">Blocks</a></li><li><a class="tocitem" href="lib/edge/">Edges</a></li><li><a class="tocitem" href="lib/graph/">Graphs</a></li><li><a class="tocitem" href="lib/gfa/">GFA</a></li><li><a class="tocitem" href="lib/mash/">Mash Implementation</a></li><li><a class="tocitem" href="lib/minimap/">Minimap2 Wrapper</a></li><li><a class="tocitem" href="lib/node/">Nodes</a></li><li><a class="tocitem" href="lib/path/">Paths</a></li><li><a class="tocitem" href="lib/simulate/">Simulation</a></li><li><a class="tocitem" href="lib/utility/">Utility</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="cli/build/">Build</a></li><li><a class="tocitem" href="cli/export/">Export</a></li><li><a class="tocitem" href="cli/generate/">Generate</a></li><li><a class="tocitem" href="cli/marginalize/">Marginalize</a></li><li><a class="tocitem" href="cli/polish/">Polish</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/nnoll/pangraph/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="PanGraph"><a class="docs-heading-anchor" href="#PanGraph">PanGraph</a><a id="PanGraph-1"></a><a class="docs-heading-anchor-permalink" href="#PanGraph" title="Permalink"></a></h1><p>A <strong>fast</strong>, <strong>self-contained</strong> Julia library and command line tool suite that aligns multiple genomes simultaneously.</p><h2 id="Introduction"><a class="docs-heading-anchor" href="#Introduction">Introduction</a><a id="Introduction-1"></a><a class="docs-heading-anchor-permalink" href="#Introduction" title="Permalink"></a></h2><p>Microbes don&#39;t evolve strictly by vertical descent and modification; in addition, there is persistent lateral exchange of genetic material within their local spatial neighborhood. At the present, explanatory population genetic models exist at both extrema of limits horizontal transfer rate; researchers have a reasonable, quantitative understanding of both asexual and the fully decoupled, single-site, evolution. However, there are currently no satisfactory models, and thus computational tools, in the more realistic intermediate regime, let alone that models which account for <em>any</em> structural variation within a population. In order to investigate such questions empirically in natural populations, we have developed a scalable multiple genome alignment tool, <strong>PanGraph</strong>, that identifies regions of mutual homology between large sets of closely related genomes. This is expected to be useful to parsimoniously infer horizontal gene transfer events within a community; perform comparative studies of genome gain, loss, and rearrangement dynamics; or simply to compress many related genomes.</p><p><strong>PanGraph</strong> (short for pangenome graph) builds a graph-based coordinate system for polymorphic variation in a microbial population that generalizes classical linear alignment techniques. The graph is built by progressively aligning pairs of genomes using Minimap2&#39;s minimizer-based algorithm, and thus simultaneously aligns all genomes within the population in linear time. The resultant graph represents contiguous intervals of homologous DNA as vertices and every genome as an ordered walk across such vertices. Edges of the graph are unordered and only exist if at least one genome was found to connect both vertices in either the forward or reverse strand. The documentation, and source code, uses the following terminology:</p><ol><li><strong>Block/PanContig:</strong>  A contiguous interval of sequence that compresses a linear multiple sequence alignment and represents a single vertex within a pangraph.  It is summarized by a consensus sequence, along with the polymorphisms needed for each genome to reconstruct the true sequence.</li><li><strong>Node:</strong>  An oriented traversal of a genome through a <em>block</em>, i.e. whether a given genome/duplication traverses a block along the forward or reverse direction.  One genome can have multiple nodes that point to the same <em>block</em>.</li><li><strong>Path:</strong>  A sequence of <em>nodes</em> that represent an ordered walk on the graph that reconstructs an input genome.  Can be circular or linear.</li><li><strong>Edge:</strong>  A juxtaposition/breakpoint between two <em>blocks</em> found in at least one genome.</li><li><strong>Graph/PanGraph:</strong>  A collection of <em>blocks</em>, associated to all recognized intervals of homology, and <em>paths</em>, genomes stored as an ordered walk of <em>nodes</em>.</li></ol><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>There are multiple ways to install PanGraph</p><h3 id="From-Julia-REPL"><a class="docs-heading-anchor" href="#From-Julia-REPL">From Julia REPL</a><a id="From-Julia-REPL-1"></a><a class="docs-heading-anchor-permalink" href="#From-Julia-REPL" title="Permalink"></a></h3><pre><code class="language-julia hljs">    (@v1.x) pkg&gt; add https://github.com/nnoll/pangraph.git</code></pre><h3 id="From-Command-Line"><a class="docs-heading-anchor" href="#From-Command-Line">From Command Line</a><a id="From-Command-Line-1"></a><a class="docs-heading-anchor-permalink" href="#From-Command-Line" title="Permalink"></a></h3><pre><code class="language-bash hljs">    julia -e &#39;using Pkg; Pkg.add(&quot;https://github.com/nnoll/pangraph.git&quot;)&#39;</code></pre><h3 id="Local-Environment"><a class="docs-heading-anchor" href="#Local-Environment">Local Environment</a><a id="Local-Environment-1"></a><a class="docs-heading-anchor-permalink" href="#Local-Environment" title="Permalink"></a></h3><p>Clone the repository.</p><pre><code class="language-bash hljs">    git clone https://github.com/nnoll/pangraph.git &amp;&amp; cd pangraph</code></pre><p>Instantiate the package. This will create a seperate Julia environment for PanGraph</p><pre><code class="language-bash hljs">    julia -e &#39;using Pkg; Pkg.instantiate(&quot;.&quot;)&#39;</code></pre><p>Enter the REPL</p><pre><code class="language-bash hljs">    julia --project=.</code></pre><h3 id="Optional-dependencies"><a class="docs-heading-anchor" href="#Optional-dependencies">Optional dependencies</a><a id="Optional-dependencies-1"></a><a class="docs-heading-anchor-permalink" href="#Optional-dependencies" title="Permalink"></a></h3><p>There are a few <strong>optional</strong> external programs that PanGraph can utilize</p><ol><li><a href="https://github.com/marbl/Mash">Mash</a> can be used to construct a guide tree in place of our internal algorithm.</li><li><a href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> can be optionally used to polish homologous alignments. Only recommended for short alignments.</li></ol><p>In order to invoke functionality from PanGraph, these tools must be installed and available on <strong>PATH</strong></p><h2 id="User&#39;s-Guide"><a class="docs-heading-anchor" href="#User&#39;s-Guide">User&#39;s Guide</a><a id="User&#39;s-Guide-1"></a><a class="docs-heading-anchor-permalink" href="#User&#39;s-Guide" title="Permalink"></a></h2><p>Basic functionality of PanGraph is provided by a command line interface. This includes multiple genome alignment, the export of a genome alignment to various visualization formats, alignment polishing, and genome comparison tool. Additionally, generation of basic synthetic data is included for testing.</p><p>For uncovered use cases, functionality can be added by utilizing the underlying library functions. Please see the high-level overview for definitions of library terminology.</p><h2 id="Citing-PanGraph"><a class="docs-heading-anchor" href="#Citing-PanGraph">Citing PanGraph</a><a id="Citing-PanGraph-1"></a><a class="docs-heading-anchor-permalink" href="#Citing-PanGraph" title="Permalink"></a></h2><p>Add citation here</p></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="lib/align/">Alignment »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.10 on <span class="colophon-date" title="Monday 29 November 2021 16:12">Monday 29 November 2021</span>. Using Julia version 1.6.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
